lib/vault/api/logical.rb

```
@@ -1,150 +1,28 @@
-      json[:data][:keys] || []
-    rescue HTTPError => e
-      return [] if e.code == 404
-      raise
-    end
-
-    # Read the secret at the given path. If the secret does not exist, +nil+
-    # will be returned.
-    #
-    # @example
-    #   Vault.logical.read("secret/password") #=> #<Vault::Secret lease_id="">
-    #
-    # @param [String] path
-    #   the path to read
-    #
-    # @return [Secret, nil]
-    def read(path, options = {})
-      headers = extract_headers!(options)
-      json = client.get("/v1/#{encode_path(path)}", {}, headers)
-      return Secret.decode(json)
-    rescue HTTPError => e
-      return nil if e.code == 404
-      raise
-    end
-
-    # Write the secret at the given path with the given data. Note that the
-    # data must be a {Hash}!
-    #
-    # @example
-    #   Vault.logical.write("secret/password", value: "secret") #=> #<Vault::Secret lease_id="">
-    #
-    # @param [String] path
-    #   the path to write
-    # @param [Hash] data
-    #   the data to write
-    #
-    # @return [Secret]
-    def write(path, data = {}, options = {})
-      headers = extract_headers!(options)
-      json = client.put("/v1/#{encode_path(path)}", JSON.fast_generate(data), headers)
-      if json.nil?
-        return true
-      else
-        return Secret.decode(json)
-      end
-    end
+    LOGICAL_APPROACHES = {
+      unversioned: Logical::Unversioned,
+      versioned:   Logical::Versioned
+    }.freeze
 
-    # Delete the secret at the given path. If the secret does not exist, vault
-    # will still return true.
-    #
-    # @example
-    #   Vault.logical.delete("secret/password") #=> true
-    #
-    # @param [String] path
-    #   the path to delete
-    #
-    # @return [true]
-    def delete(path)
-      client.delete("/v1/#{encode_path(path)}")
-      return true
+    # A proxy to the {Logical} methods.
+    # @return [Logical::Unversioned, Logical::Versioned]
+    def logical(approach = :unversioned)
```

might be better to require the argument to avoid confusion ?
also `logicals.fetch(approach)` to fail fast

---

That'd be a breaking change, which I'm hesitant to do. At least this way it should be backwards compatible, and then future versions can mark an unprovided argument as deprecated, and then remove the old approach altogether.